### ISSUE TITLE
Problem of group_keypoints

### ISSUE THREAD

**Issue by yoletPig** [Daniil-Osokin/lightweight-human-pose-estimation.pytorch#306]
@Daniil-Osokin Thanks for your great job! I need some help for training custom dataset.I have trained a horse dataset with 22 keypoints and I set 10 pairs of paf.I show the heatmap of each paf and the keypoints which is predicted by iter_19500.pth.I set the min_paf_score is 0 and the success_ratio is 0.1 but some limbs still failed.
![image](https://user-images.githubusercontent.com/49870725/226390829-ea7533d7-2c70-40dd-9410-64141f0b398a.png)
![image](https://user-images.githubusercontent.com/49870725/226390896-a15a7e16-0c9e-476d-b4c3-4d5d97bee1f4.png)
![image](https://user-images.githubusercontent.com/49870725/226390932-50b8011b-0445-4ea0-ba02-f5e74eac7ea1.png)
![image](https://user-images.githubusercontent.com/49870725/226390954-38900fae-bad0-478e-be3c-fc643319adb7.png)
![image](https://user-images.githubusercontent.com/49870725/226390999-957bba8c-4e14-4ba4-a260-b147a1e1532a.png)
![image](https://user-images.githubusercontent.com/49870725/226391020-34344c0c-d212-4da0-ab0a-6806ccbb0e90.png)
![image](https://user-images.githubusercontent.com/49870725/226391043-dbb4330d-ce56-47f7-9dbf-45edd1d406b4.png)
![image](https://user-images.githubusercontent.com/49870725/226391086-335aa4f0-d03a-4f03-b9e5-ac39ed87b8c4.png)
![image](https://user-images.githubusercontent.com/49870725/226391126-4159a171-51c6-4357-b141-5a4b60fd6a44.png)
![image](https://user-images.githubusercontent.com/49870725/226391161-0cb5093c-64ad-46ea-976c-84f03519a878.png)
![image](https://user-images.githubusercontent.com/49870725/226391213-4c5357d6-37af-4cdb-a338-2b52b4a89b60.png)
![image](https://user-images.githubusercontent.com/49870725/226391251-c42f75dc-6471-4d91-9861-7442072b9449.png)
![image](https://user-images.githubusercontent.com/49870725/226391279-c10dbaf3-8ff3-4463-acbd-59de93bfc2c3.png)
![image](https://user-images.githubusercontent.com/49870725/226391334-26f4fd84-4af5-46e9-a8de-c4e60e9b8f69.png)
![image](https://user-images.githubusercontent.com/49870725/226391363-92dde48f-4e35-4ad6-9301-7f673c57b7a1.png)
![image](https://user-images.githubusercontent.com/49870725/226391382-6bbf16bc-8ee4-4333-b440-548946e6d48b.png)
![image](https://user-images.githubusercontent.com/49870725/226391408-130ed2e4-bd50-4149-b6ba-2aa519ccef88.png)
![image](https://user-images.githubusercontent.com/49870725/226391434-d1af835f-ec94-41ce-b3fa-f4f33a596d2f.png)
![image](https://user-images.githubusercontent.com/49870725/226391466-76c2b679-96d7-4ee1-b4c8-0eeea099d599.png)
![image](https://user-images.githubusercontent.com/49870725/226391499-f3402ede-3183-49c8-949a-422bc75f5b67.png)
![image](https://user-images.githubusercontent.com/49870725/226391806-3a892719-08e9-4fc0-9bbd-0867db355b6f.png)
![image](https://user-images.githubusercontent.com/49870725/226393570-b2f35e60-7cb4-42ab-8019-814b84c701bf.png)

**Comment by Daniil-Osokin** [comment]
Hi! I did not get which pafs you are using exactly (between which keypoints). However, I believe that it might be hard for the network to learn paf over background, e.g. in the last image, the horse ankle is connected with horse hoof, and so on (in this example paf goes over background). This might be a reason that grouping of some parts fail.

**Comment by Daniil-Osokin** [comment]
Hope, it is helped.

**Comment by yoletPig** [comment]
I'm so sorry to reply you so late,I have solved this problem,I found the setting of pafs are error by debug. I only set the horse ankle connect to the hoof before for reducing the limbs connection.

**Comment by Daniil-Osokin** [comment]
Great to hear it, good luck!

**Comment by siewwentong** [comment]
@yoletPig can you share the code you used to display the heatmaps and paf, would really appreciate that, thanks!

**Comment by yoletPig** [comment]
my test code
`import cv2
import numpy as np
import torch
from matplotlib import pyplot as plt
from torch.utils.data import DataLoader
from torchvision import transforms

from openpose.datasets.coco import CocoTrainDataset
from openpose.datasets.transformations import Scale, Rotate, CropPad
from openpose.models.with_mobilenet import PoseEstimationWithMobileNet
from openpose.modules.load_state import load_state
from openpose.val import infer, heatmap2d

def heatmap2d(arr: np.ndarray, image_constant, id_of_paf):
    plt.imshow(image_constant)
    plt.imshow(arr, cmap='viridis', alpha=0.6)
    plt.title(str(id_of_paf))
    plt.colorbar()
    plt.show()

net = PoseEstimationWithMobileNet(num_refinement_stages=3, num_heatmaps=23, num_pafs=20)
checkpoint = torch.load("default_checkpoints/checkpoint_iter_11000.pth")
load_state(net, checkpoint)
net.cuda()
prepared_train_labels = "prepared_train_annotation.pkl"
train_images_folder = "../../mmpose_data/Horse-10/"
stride = 8
sigma = 7
path_thickness = 1
dataset = CocoTrainDataset(prepared_train_labels, train_images_folder,
                           stride, sigma, path_thickness,
                           transform=transforms.Compose([
                               Scale(),
                               CropPad(pad=(128, 128, 128))
                           ])
                           )
train_loader = DataLoader(dataset, batch_size=32, shuffle=False, num_workers=8)
plot_index = 0
for batch_data in train_loader:
    plot_index += 1
    if plot_index < 20:
        continue
    images = batch_data['image'].cuda()
    keypoint_masks = batch_data['keypoint_mask'].cuda()
    paf_masks = batch_data['paf_mask'].cuda()
    keypoint_maps = batch_data['keypoint_maps'].cuda()
    paf_maps = batch_data['paf_maps'].cuda()
    index = 31
    image = images[index].detach().cpu().numpy()
    image = image.transpose((1, 2, 0))
    image = image * 256 + 128
    # avg_heatmaps, avg_pafs, pafs = infer(net, image, scales=[1], base_height=368, stride=stride)
    # image = images[index].detach().cpu().numpy()
    # image = image.transpose((1, 2, 0))
    # image = image * 256 + 128
    paf_map = paf_maps[index].detach().cpu().numpy()
    keypoint_map = keypoint_maps[index].detach().cpu().numpy()
    # paf_map = cv2.resize(paf_map[19], dsize=None, fx=stride, fy=stride, interpolation=0)
    # image[:, :, 0] = image[:, :, 0] * (paf_map*5 + 1)
    # image[:, :, 1] = image[:, :, 1] * (paf_map*5 + 1)
    # image[:, :, 2] = image[:, :, 2] * (paf_map*5 + 1)
    # cv2.imwrite("Figure.jpg", image)
    # paf_map = avg_pafs[:, :, 1]
    # paf_map = np.logical_and(paf_map, paf_map > 0).astype(np.int8)
    # paf_map = avg_pafs[:, :, 19] * paf_map
    # image[:, :, 0] = image[:, :, 0] * (paf_map * 10 + 1)
    # image[:, :, 1] = image[:, :, 1] * (paf_map * 10 + 1)
    # image[:, :, 2] = image[:, :, 2] * (paf_map * 10 + 1)
    # cv2.imwrite("Figure.jpg", image)
    image = image.astype(int)
    for idx_paf in range(20):
        paf_map_i = cv2.resize(paf_map[idx_paf], dsize=None, fx=stride, fy=stride, interpolation=0)
        heatmap2d(paf_map_i, image, idx_paf)

    htmp = np.sum(keypoint_map[:22,:,:], axis=0)
    pafm = np.sum(paf_map, axis=0)
    htmp = cv2.resize(htmp, dsize=image.shape[:2],  interpolation=0)
    pafm = cv2.resize(pafm, dsize=image.shape[:2],  interpolation=0)
    fig = plt.figure(figsize=(8, 6))
    ax1 = fig.add_subplot(211)
    ax2 = fig.add_subplot(212)
    ax1.title.set_text('Heatmap ' + str(1))
    ax1.imshow(htmp, cmap='RdYlBu')
    im1 = ax1.imshow(image, alpha=0.4)

    fig.colorbar(im1, ax=ax1, orientation='vertical')

    ax2.title.set_text('PAF' + str(1))
    ax2.imshow(pafm, cmap='RdYlBu')
    im2 = ax2.imshow(image, alpha=0.4)

    fig.colorbar(im2, ax=ax2, orientation='vertical')
    plt.show()
    print()
    break
`

**Comment by xin4431** [comment]
Hello, I have some questions about training custom datasets, can you price contact information?
Or answer online

**Comment by xin4431** [comment]
I'm currently doing pose tests on pigs, using 20 key points. The following problems exist:
1. The mapping between BODY_PARTS_KPT_IDS and BODY_PARTS_PAF_IDS is not clear. I'd like to ask you something.
2. I modified part of the code according to my own data set in accordance with the training process. Then run 
`python train.py --train-images-folder /train2017/ -- preparation-train-labels prepared_train_annotation.pkl --val-labels val_subset.json --val-images-folder /val2017/ --checkpoint-path /mobilenet_sgd_68.848.pth.tar --from-mobilenet`

 When    (train.py  132#)
`evaluating (val_labels, val_output_name, val_images_folder, net), avg_heatmaps, avg_pafs = infer(net, img, scales, base_height,`

The value in avg_heatmaps is very small, close to 0.0001
In` total_keypoints_num += extract_keypoints(avg_heatmaps[:, :, kpt_idx], all_keypoints_by_type, total_keypoints_num)`
After heatmap[heatmap < 0.1] = 0, the values of avg_heatmaps are all zeros.

Eventually, the `detections.json` file is empty, resulting in an error like [#304](https://github.com/Daniil-Osokin/lightweight-human-pose-estimation.pytorch/issues/304)!
求大佬指教啊

**Comment by yoletPig** [comment]
你要修改coco eval方法里面的sigma值，默认只有17个关键点

**Comment by yoletPig** [comment]
> 

你可以加我微信13025811310
